packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx

```diff
@@ -69,106 +67,184 @@ export const ModelsTreeComponent = (props: ModelTreeProps) => {
       });
   }, [queryModels, viewport]);
 
-  const invert = useCallback(async () => {
-    if (availableModels.length === 0) return;
-    const notViewedModels: string[] = [];
-    const models: string[] = [];
-    availableModels.forEach((id: string) => {
-      if (viewport?.viewsModel(id)) models.push(id);
-      else notViewedModels.push(id);
-    });
-    await viewport?.addViewedModels(notViewedModels);
-    viewport?.changeModelDisplay(models, false);
-    viewport?.invalidateScene();
-  }, [viewport, availableModels]);
+  return (
+    <>
+      {iModel && viewport &&
+        <>
+          <TreeHeaderComponent
+            searchOptions={searchOptions}
+            treeHeaderButtons={props.TreeHeaderButtons
+              ? props.TreeHeaderButtons.map((btn) => btn({ viewport, iModel, availableModels }))
+              : [
+                <ShowAllButtonModelsTree viewport={viewport} availableModels={availableModels} iModel={iModel} key="show-all-btn" />,
+                <HideAllButtonModelsTree viewport={viewport} availableModels={availableModels} iModel={iModel} key="hide-all-btn" />,
+                <InvertButtonModelsTree viewport={viewport} availableModels={availableModels} iModel={iModel} key="invert-all-btn" />,
+                <View2DButtonModelsTree viewport={viewport} availableModels={available2dModels} key="view-2d-btn" />,
+                <View3DButtonModelsTree viewport={viewport} availableModels={available3dModels} key="view-3d-btn" />,
+              ]
+            }
+          />
+          <AutoSizer>
+            {({ width, height }) => (
+              <ModelsTree
+                {...props}
+                iModel={iModel}
+                activeView={viewport}
+                width={width}
+                height={height}
+                filterInfo={{ filter: filterString, activeMatchIndex }}
+                onFilterApplied={onFilterApplied}
+              />
+            )}
+          </AutoSizer>
+        </>
+      }
+    </>
+  );
+};
 
-  const hideAll = useCallback(async () => {
-    viewport?.changeModelDisplay(availableModels, false);
-    viewport?.clearAlwaysDrawn();
-    if (iModel) {
+export function ShowAllButtonModelsTree(props: ModelsTreeHeaderButtonProps) {
+  const showAll = useCallback(async () => {
```

Comment:
I think there is no need to wrap `onClick` handler with `useCallback` in all of these buttons

Reply:
fixed